pkg/broker/receiver.go

```diff
@@ -0,0 +1,141 @@
+	if err := r.initClient(); err != nil {
+		r.logger.Warn("Failed to initialize client", zap.Error(err))
+	}
+	return provisioners.NewMessageReceiver(r.sendEvent, r.logger.Sugar())
+}
+
+// sendEvent sends an event to a subscriber if the trigger filter passes.
+func (r *Receiver) sendEvent(trigger provisioners.ChannelReference, message *provisioners.Message) error {
+	r.logger.Debug("Received message", zap.Any("triggerRef", trigger))
+	ctx := context.Background()
+
+	t, err := r.getTrigger(ctx, trigger)
+	if err != nil {
+		r.logger.Info("Unable to get the Trigger", zap.Error(err), zap.Any("triggerRef", trigger))
+		return err
+	}
+
+	subscriberURI := t.Status.SubscriberURI
+	if subscriberURI == "" {
+		r.logger.Error("Unable to read subscriberURI")
+		return errors.New("unable to read subscriberURI")
+	}
+
+	if !r.shouldSendMessage(&t.Spec, message) {
+		r.logger.Debug("Message did not pass filter", zap.Any("triggerRef", trigger))
+		return nil
+	}
+
+	err = r.dispatcher.DispatchMessage(message, subscriberURI, "", provisioners.DispatchDefaults{})
+	if err != nil {
+		r.logger.Info("Failed to dispatch message", zap.Error(err), zap.Any("triggerRef", trigger))
+		return err
+	}
+	r.logger.Debug("Successfully sent message", zap.Any("triggerRef", trigger))
+	return nil
+}
+
+// Initialize the client. Mainly intended to load stuff in its cache.
+func (r *Receiver) initClient() error {
+	// We list triggers so that we can load the client's cache. Otherwise, on receiving an event, it
+	// may not find the trigger and would return an error.
+	opts := &client.ListOptions{
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		tl := &eventingv1alpha1.TriggerList{}
+		if err := r.client.List(context.TODO(), opts, tl); err != nil {
+			return err
+		}
+		if tl.Continue != "" {
+			opts.Raw.Continue = tl.Continue
+		} else {
+			break
+		}
+	}
+	return nil
+}
+
+func (r *Receiver) getTrigger(ctx context.Context, ref provisioners.ChannelReference) (*eventingv1alpha1.Trigger, error) {
+	t := &eventingv1alpha1.Trigger{}
+	err := r.client.Get(ctx,
+		types.NamespacedName{
+			Namespace: ref.Namespace,
+			Name:      ref.Name,
+		},
+		t)
+	return t, err
+}
+
+// shouldSendMessage determines whether message 'm' should be sent based on the triggerSpec 'ts'.
+// Currently it supports exact matching on type and/or source of events.
+func (r *Receiver) shouldSendMessage(ts *eventingv1alpha1.TriggerSpec, m *provisioners.Message) bool {
+	if ts.Filter == nil || ts.Filter.SourceAndType == nil {
+		r.logger.Error("No filter specified")
+		return false
+	}
+	filterType := ts.Filter.SourceAndType.Type
+	if filterType != eventingv1alpha1.TriggerAnyFilter && filterType != m.Headers["Ce-Eventtype"] {
+		r.logger.Debug("Wrong type", zap.String("trigger.spec.filter.sourceAndType.type", filterType), zap.String("message.type", m.Headers["Ce-Eventtype"]))
```

Comment:
nit. Type mismatch rather than wrong type?
Sam comment for Source mismatch.

Reply:
@n3wscott  will add the SDK for 0.5, so this will change later on. But is not blocking for the PR.

Reply:
@Harwayne will also add the Ce-Type header for now. Source I think is called the same (Ce-Source) in both versions of the spec, but we should double check.